pgjdbc/src/test/java/org/postgresql/test/util/PasswordUtilTest.java

```diff
@@ -0,0 +1,159 @@
+      }
+    });
+  }
+
+  private void testUserPassword(@Nullable String encryptionType, String username, String password,
+      String encodedPassword) throws SQLException {
+    String escapedUsername = Utils.escapeIdentifier(null, username).toString();
+
+    try (Connection superConn = TestUtil.openPrivilegedDB()) {
+      TestUtil.execute(superConn, "CREATE USER " //
+          + escapedUsername //
+          + " WITH PASSWORD '" + encodedPassword + "'");
+
+      String shadowPass = TestUtil.queryForString(superConn, //
+          "SELECT passwd FROM pg_shadow WHERE usename = ?", username);
+      Assert.assertEquals("pg_shadow value of password must match encoded", shadowPass, encodedPassword);
+
+      // We should be able to log in using our new user:
+      assertValidUsernamePassword(username, password);
+      // We also check that we cannot log in with the wrong password to ensure that
+      // the server is not simply trusting everything
+      assertInvalidUsernamePassword(username, "Bad Password:" + password);
+
+      String newPassword = "mySecretNewPassword" + randomSuffix();
+      PGConnection pgConn = superConn.unwrap(PGConnection.class);
+      pgConn.alterUserPassword(username, () -> newPassword.toCharArray(), encryptionType);
+
+      // We should be able to log in using our new password
+      assertValidUsernamePassword(username, newPassword);
+      // We also check that we cannot log in with the wrong password to ensure that
+      // the server is not simply trusting everything
+      assertInvalidUsernamePassword(username, "Bad Password:" + newPassword);
+    } finally {
+      try (Connection superConn = TestUtil.openPrivilegedDB()) {
+        TestUtil.execute(superConn, "DROP USER " + escapedUsername);
+      } catch (Exception ignore) { }
+    }
+  }
+
+  private void testUserPassword(@Nullable String encryptionType, String username, String password) throws SQLException {
+    String encodedPassword = PasswordUtil.encodePassword(username, () -> password.toCharArray(), encryptionType);
+    testUserPassword(encryptionType, username, password, encodedPassword);
+  }
+
+  private void testUserPassword(@Nullable String encryptionType) throws SQLException {
+    String username = "test_password_" + randomSuffix();
+    String password = "t0pSecret" + randomSuffix();
+
+    testUserPassword(encryptionType, username, password);
+    testUserPassword(encryptionType, username, "password with spaces");
+    testUserPassword(encryptionType, username, "password with single ' quote'");
+    testUserPassword(encryptionType, username, "password with double \" quote'");
+    testUserPassword(encryptionType, username + " with spaces", password);
+    testUserPassword(encryptionType, username + " with single ' quote", password);
+    testUserPassword(encryptionType, username + " with single \" quote", password);
+  }
+
+  @Test
+  public void testServerDefault() throws SQLException {
```

Comment:
Technically speaking, JUnit4 does not use `test..` method naming. `test...` method prefix comes from JUnit3 when `@Test` annotation did not exist.

Reply:
I still like it though as it makes the method name an imperative mood. Without the prefix we get `public void serverDefault()` which sounds like it returns a value. Could be more clever and do something like "validateServerDefault" but meh...

Reply:
`testServerDefault` does not explain what the test verifies though